libc/src/time/time_utils.h

```diff
@@ -94,6 +101,255 @@ LIBC_INLINE struct tm *localtime(const time_t *t_ptr) {
+    if (days_until_end_of_year < 3)
+      return 1;
+
+    // else just calculate the current week like normal.
+    const int ceil_weeks_since_start =
+        ((start_of_cur_week + 1) + (time_constants::DAYS_PER_WEEK - 1)) /
+        time_constants::DAYS_PER_WEEK;
+
+    // add 1 if this year's first week starts in the previous year.
+    return ceil_weeks_since_start +
+           (((start_of_cur_week + time_constants::DAYS_PER_WEEK) %
+             time_constants::DAYS_PER_WEEK) >
+                    time_constants::ISO_FIRST_DAY_OF_YEAR
+                ? 1
+                : 0);
+  }
+
+  LIBC_INLINE constexpr int get_iso_year() const {
+    const int BASE_YEAR = get_year();
+    // The ISO year is the same as a standard year for all dates after the start
+    // of the first week and before the last week. Since the first ISO week of a
+    // year starts on the 4th, anything after that is in this year.
+    if (timeptr->tm_yday >= time_constants::ISO_FIRST_DAY_OF_YEAR &&
+        timeptr->tm_yday < time_constants::DAYS_PER_NON_LEAP_YEAR -
+                               time_constants::DAYS_PER_WEEK)
+      return BASE_YEAR;
+
+    const int ISO_WDAY = get_iso_wday();
+    // The first week of the ISO year is defined as the week containing the
+    // 4th day of January.
+
+    // first week
+    if (timeptr->tm_yday < time_constants::ISO_FIRST_DAY_OF_YEAR) {
+      /*
+      If jan 4 is in this week, then we're in BASE_YEAR, else we're in the
+      previous year. The formula's been rearranged so here's the derivation:
+
+              +--------+-- days until jan 4
+              |        |
+       wday + (4 - yday) < 7
+       |               |
+       +---------------+-- weekday of jan 4
+
+       rearranged to get all the constants on one side:
+
+       wday - yday < 7 - 4
+      */
+      return (ISO_WDAY - timeptr->tm_yday <
+              time_constants::DAYS_PER_WEEK -
+                  time_constants::ISO_FIRST_DAY_OF_YEAR)
+                 ? BASE_YEAR
+                 : BASE_YEAR - 1;
+    }
+
+    // last week
+    const int DAYS_LEFT_IN_YEAR =
+        get_days_in_year(get_year()) - timeptr->tm_yday;
+    /*
+    Similar to above, we're checking if jan 4 (of next year) is in this week. If
+    it is, this is in the next year. Note that this also handles the case of
+    yday > days in year gracefully.
+
+           +------------------+-- days until jan 4 (of next year)
+           |                  |
+    wday + (4 + remaining days) < 7
+    |                         |
+    +-------------------------+-- weekday of jan 4
+
+    rearranging we get:
+
+    wday + remaining days < 7 - 4
+    */
+
+    return (ISO_WDAY + DAYS_LEFT_IN_YEAR <
+            time_constants::DAYS_PER_WEEK -
+                time_constants::ISO_FIRST_DAY_OF_YEAR)
+               ? BASE_YEAR + 1
+               : BASE_YEAR;
+  }
+
+  LIBC_INLINE constexpr time_t get_epoch() const {
+    return mktime_internal(timeptr);
```

Comment:
If the member `timeptr` was modified, is calling `mktime_internal` with it guaranteed to return epoch?